pkg/runtime/processor/binding.go

```diff
@@ -0,0 +1,135 @@
+	compapi "github.com/dapr/dapr/pkg/apis/components/v1alpha1"
+	compbindings "github.com/dapr/dapr/pkg/components/bindings"
+	diag "github.com/dapr/dapr/pkg/diagnostics"
+	"github.com/dapr/dapr/pkg/runtime/compstore"
+	rterrors "github.com/dapr/dapr/pkg/runtime/errors"
+	"github.com/dapr/dapr/pkg/runtime/meta"
+)
+
+const (
+	bindingTypeInput  = "input"
+	bindingTypeOutput = "output"
+	bindingDirection  = "direction"
+)
+
+type binding struct {
+	registry  *compbindings.Registry
+	compStore *compstore.ComponentStore
+	meta      *meta.Meta
+}
+
+func (b *binding) init(ctx context.Context, comp compapi.Component) error {
+	switch {
+	case b.registry.HasInputBinding(comp.Spec.Type, comp.Spec.Version):
+		if err := b.initInputBinding(ctx, comp); err != nil {
+			return err
+		}
+
+	case b.registry.HasOutputBinding(comp.Spec.Type, comp.Spec.Version):
+		if err := b.initOutputBinding(ctx, comp); err != nil {
+			return err
+		}
+
+	default:
+		diag.DefaultMonitoring.ComponentInitFailed(comp.Spec.Type, "creation", comp.ObjectMeta.Name)
+		return fmt.Errorf("couldn't find binding %s/%s", comp.Spec.Type, comp.Spec.Version)
+	}
+
+	return nil
+}
+
+func (b *binding) initInputBinding(ctx context.Context, comp compapi.Component) error {
+	if !b.isBindingOfDirection(bindingTypeInput, comp.Spec.Metadata) {
+		return nil
+	}
+
+	fName := comp.LogName()
+	binding, err := b.registry.CreateInputBinding(comp.Spec.Type, comp.Spec.Version, fName)
+	if err != nil {
+		diag.DefaultMonitoring.ComponentInitFailed(comp.Spec.Type, "creation", comp.ObjectMeta.Name)
+		return rterrors.NewInit(rterrors.CreateComponentFailure, fName, err)
+	}
+	err = binding.Init(ctx, bindings.Metadata{Base: b.meta.ToBaseMetadata(comp)})
+	if err != nil {
+		diag.DefaultMonitoring.ComponentInitFailed(comp.Spec.Type, "init", comp.ObjectMeta.Name)
+		return rterrors.NewInit(rterrors.InitComponentFailure, fName, err)
+	}
+
+	log.Infof("successful init for input binding %s (%s/%s)", comp.ObjectMeta.Name, comp.Spec.Type, comp.Spec.Version)
+	b.compStore.AddInputBindingRoute(comp.Name, comp.Name)
+	for _, item := range comp.Spec.Metadata {
+		if item.Name == "route" {
+			b.compStore.AddInputBindingRoute(comp.ObjectMeta.Name, item.Value.String())
+			break
+		}
+	}
+	b.compStore.AddInputBinding(comp.Name, binding)
+	diag.DefaultMonitoring.ComponentInitialized(comp.Spec.Type)
+	return nil
+}
+
+func (b *binding) initOutputBinding(ctx context.Context, comp compapi.Component) error {
+	if !b.isBindingOfDirection(bindingTypeOutput, comp.Spec.Metadata) {
+		return nil
+	}
+
+	fName := comp.LogName()
+	binding, err := b.registry.CreateOutputBinding(comp.Spec.Type, comp.Spec.Version, fName)
+	if err != nil {
+		diag.DefaultMonitoring.ComponentInitFailed(comp.Spec.Type, "creation", comp.ObjectMeta.Name)
+		return rterrors.NewInit(rterrors.CreateComponentFailure, fName, err)
+	}
+
+	if binding != nil {
+		err := binding.Init(context.TODO(), bindings.Metadata{Base: b.meta.ToBaseMetadata(comp)})
```

Comment:
Was this supposed to be `ctx`?

```suggestion
		err := binding.Init(ctx, bindings.Metadata{Base: b.meta.ToBaseMetadata(comp)})
```

Reply:
It needs to be `ctx` eventually yes. Context propagation is part of this runtime refactoring effort to eventually achieve graceful shutdown + hot reloading.